plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,633 @@
+
+		if (!g_usb_device_control_transfer(usb_device,
+						   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+						   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+						   G_USB_DEVICE_RECIPIENT_INTERFACE,
+						   ALGOLTEK_USB_REQUEST_CMD06,
+						   0, /* value */
+						   0, /* idx */
+						   partIspData->data,
+						   partIspData->data[0],
+						   NULL,
+						   ALGOLTEK_DEVICE_USB_TIMEOUT,
+						   NULL,
+						   error)) {
+			g_prefix_error(error, "failed to contact device: ");
+			return FALSE;
+		}
+
+		startIndex += transferDataSize;
+		restTransferDataSize -= transferDataSize;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_device_CMD07(FuAlgoltekUsbDevice *self, int address, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	fu_byte_array_set_size(buf, 11, 0x0);
+	buf->data[0] = 5;
+	buf->data[1] = ALGOLTEK_USB_REQUEST_CMD07;
+	buf->data[2] = ((address >> 8) & 0xFF);
+	buf->data[3] = (address & 0xFF);
+	buf->data[10] = checkSum(buf);
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   ALGOLTEK_USB_REQUEST_CMD07,
+					   0, /* value */
+					   0, /* idx */
+					   buf->data,
+					   buf->data[0],
+					   NULL,
+					   ALGOLTEK_DEVICE_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "failed to contact device: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_device_CMD08(FuAlgoltekUsbDevice *self, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	fu_byte_array_set_size(buf, 2, 0x0);
+	buf->data[0] = 0;
+	buf->data[1] = 0;
+
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   ALGOLTEK_USB_REQUEST_CMD08,
+					   0, /* value */
+					   0, /* idx */
+					   buf->data,
+					   buf->len,
+					   NULL,
+					   ALGOLTEK_DEVICE_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "failed to contact device: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_device_status_check(FuAlgoltekUsbDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) readoutCheck = g_byte_array_new();
+	guint retryTimes = 0;
+	guint checkResult;
+Retry:
+	readoutCheck = fu_algoltek_device_CMD02(self, 0x860C, error);
+	if (readoutCheck == NULL)
+		return FALSE;
+	checkResult = fu_algoltek_readout_info_int(readoutCheck);
+	switch (checkResult) {
+	case Result_PASS:
+		break;
+	case Result_ERROR:
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "Check flashwrite status return failed...");
+		return FALSE;
+	default:
+		retryTimes++;
+		if (retryTimes < 10)
+			goto Retry;
```

Comment:
You should use `fu_device_retry_full()` here instead of using goto.

Reply:
Updated! Use `fu_device_retry_full( )` here instead.